src/core/components/pin/gc-lock.js

```diff
@@ -0,0 +1,135 @@
+'use strict'
+
+const assert = require('assert')
+const mortice = require('mortice')
+const nanoid = require('nanoid')
+const pull = require('pull-stream/pull')
+const pullThrough = require('pull-stream/throughs/through')
+const pullAsyncMap = require('pull-stream/throughs/async-map')
+const EventEmitter = require('events')
+const log = require('debug')('ipfs:gc:lock')
+
+class GCLock extends EventEmitter {
+  constructor (repoOwner) {
+    super()
+
+    // Ensure that we get a different mutex for each instance of GCLock
+    // (There should only be one GCLock instance per IPFS instance, but
+    // there may be multiple IPFS instances, eg in unit tests)
```

Comment:
IPFS locks the repo when using it. There should never be two IPFS nodes running over the same repo. If you happen to caught it not behaving as expected, then it is a bug on the ipfs-repo module.

Reply:
@daviddias the issue here is that mortice caches the lock by name, so in order to avoid different unit tests interfering with each other we need to either make sure the name is unique, or clear the require cache for mortice before each test (debatably also a good option)

Reply:
we can also improve `mortice` to be more atomic right ?

Reply:
@achingbrain I'm not sure on the answer to this one, could you comment?

Reply:
> mortice caches the lock by name

The locks need to be accessible to the mortice module outside of the creating context in order to have the default lock work transparently across multiple cluster workers or web workers as implemented.

So if we're creating multiple in-process IPFS nodes, they'll fight over the lock name even if they use different repos.  Presumably we could make the lock name a config option or use something unique to the repo like the peer ID?